pkg/apis/tenancy/v1alpha1/types.go

```diff
@@ -102,21 +102,143 @@ type ClusterWorkspaceType struct {
+	// workspace.
+	//
 	// +optional
 	Initializers []ClusterWorkspaceInitializer `json:"initializers,omitempty"`
 
 	// additionalWorkspaceLabels are a set of labels that will be added to a
 	// ClusterWorkspace on creation.
 	//
+	// If a workspace type extends one at a higher level, both maps are merged. If they
+	// disagree about a value, the one in the sub-workspace trumps.
+	//
 	// +optional
 	AdditionalWorkspaceLabels map[string]string `json:"additionalWorkspaceLabels,omitempty"`
+
+	// defaultSubWorkspaceType is the cluster workspace type that will be used
+	// by default if another, nested ClusterWorkspace is created in a workspace
+	// of the given type. The default behaviour equals "Universal".
+	//
+	// If a workspace type extends one at a higher level and they disagree about the default
+	// the one in the sub-workspace trumps. Not specifying defaultSubWorkspaceType or
+	// an empty string means to inherit the value from the super-workspace.
+	//
+	// +optional
+	DefaultSubWorkspaceType string `json:"defaultSubWorkspaceType,omitempty"`
+
+	// allowedSubWorkspaceTypes is a list of cluster workspace types that can be
+	// created in a workspace of the given type.
+	//
+	// If a workspace type extends one at a higher level the sets of allowed sub-workspace
+	// types are merged. Sub-workspace types that are existing in a super-workspace
+	// are ignored.
+	//
+	// For example: if Organization exists at the root, a ClusterWorkspaceType called
+	// "Organization" cannot add itself to the possible allowedSubWorkspaceTypes. I.e.
+	// the nesting must be allowed at the root and cannot be overridden. I.e. the
+	// organization admin CANNOT allow further nesting of system provided workspace
+	// types.
+	//
+	// For example: if Organization exists at the root, a ClusterWorkspaceType called
+	// "Team" which does not exist at the root at root:org, can be added to the
+	// allowedSubWorkspaceTypes of the "Organization" workspace type at root:org. I.e.
+	// the organization admin can further custom workspace types to be nested under a
+	// system provided workspace type.
+	//
+	// By default no type is allowed. This means no other workspace can be nested
+	// within a workspace of the given type.
+	AllowedSubWorkspaceTypes []string `json:"allowedSubWorkspaceTypes,omitempty"`
+
+	// extensible means that the ClusterWorkspaceType can be extended in a sub-workspace.
+	//
+	// +kubebuilder:default=true
+	// +optional
+	Extensible bool `json:"extensible,omitempty"`
+}
+
+// BootstrapManifests references a set of Kubernetes manifests that are applied to a
+// workspace on initialisation.
+type BootstrapManifests struct {
+	// ConfigMaps is a list of config maps that are applied to the workspace on initialisation.
+	//
+	// +optional
+	ConfigMaps []BootstrapConfigMap `json:"configMaps,omitempty"`
+}
+
+// BootstrapConfigMap references a config map that is applied to a workspace on initialisation.
+type BootstrapConfigMap struct {
```

Comment:
Why does this not specify logical cluster?

Reply:
same workspace.